stable_baselines3/common/policies.py

```diff
@@ -228,44 +228,46 @@ def is_vectorized_observation(self, observation: Union[np.ndarray, Dict[str, np.
             )
         return vectorized_env
 
-    def obs_to_tensor(self, observation: Union[np.ndarray, Dict[str, np.ndarray]]) -> Tuple[th.Tensor, bool]:
+    def obs_to_tensor(self, observation: TorchGymObs) -> Tuple[th.Tensor, bool]:
         """
         Convert an input observation to a PyTorch tensor that can be fed to a model.
         Includes sugar-coating to handle different observations (e.g. normalizing images).
 
+        :param observation: the input observation
+        :return: The observation as PyTorch tensor
+            and whether the observation is vectorized or not
+        """
+        observation = obs_as_tensor(observation, device=self.device)
+        observation, vectorized_env = self.obs_maybe_transpose(observation)
+        return observation, vectorized_env
+
+    def obs_maybe_transpose(self, observation: TorchGymObs) -> Tuple[th.Tensor, bool]:
+        """
+        Handle different observation types (e.g. normalize images).
+
         :param observation: the input observation
         :return: The observation as PyTorch tensor
             and whether the observation is vectorized or not
         """
         vectorized_env = False
-        if isinstance(observation, dict):
-            # need to copy the dict as the dict in VecFrameStack will become a torch tensor
-            observation = copy.deepcopy(observation)
-            for key, obs in observation.items():
-                obs_space = self.observation_space.spaces[key]
-                if is_image_space(obs_space):
-                    obs_ = maybe_transpose(obs, obs_space)
-                else:
-                    obs_ = np.array(obs)
-                vectorized_env = vectorized_env or is_vectorized_observation(obs_, obs_space)
-                # Add batch dimension if needed
-                observation[key] = obs_.reshape((-1, *self.observation_space[key].shape))
-
-        elif is_image_space(self.observation_space):
-            # Handle the different cases for images
-            # as PyTorch use channel first format
-            observation = maybe_transpose(observation, self.observation_space)
 
-        else:
-            observation = np.array(observation)
+        def handle_obs(obs: th.Tensor, obs_space: spaces.Space) -> th.Tensor:
+            nonlocal vectorized_env
 
-        if not isinstance(observation, dict):
-            # Dict obs need to be handled separately
-            vectorized_env = is_vectorized_observation(observation, self.observation_space)
+            assert isinstance(obs, th.Tensor)
```

Comment:
> Is this codebase not enforcing type declarations?

Indeed code declarations are not enforced at runtime, and mypy often isn't smart enough to check all the types.

This assertion was a debugging tool (so `handle_obs` would error in a very understandable way every time the `obs` was not a Tensor) that I left there. I'll keep it there just in case :)